pkg/apis/core/helper/helpers.go

```diff
@@ -232,7 +234,8 @@ var standardQuotaResources = sets.NewString(
 // IsStandardQuotaResourceName returns true if the resource is known to
 // the quota tracking system
 func IsStandardQuotaResourceName(str string) bool {
```

Comment:
The functions in this file are used in API validation and are *really* hard to reason about changes to, especially across skewed server versions.

A feature gate is not sufficient, or the first minor version the feature gate is enabled in will allow in data the previous version of the server considers invalid.

I think the functions we're changing here have to be parameterized, and informed by the feature gate OR call sites validating updates to existing objects already containing this additional resource type.

Reply:
(it would not surprise me if this was done in an unsafe way when persistentvolumeclaim quota was added... we thought through the implications of skewed validation differences a lot more in the ~1.14 timeframe)

Reply:
I think I understand what you mean.

But suppose this goes onto 1.29, someone enables the feature gate, creates an object with a new name, and then downgrades. In 1.28 the name was not know and validation fails.

Does that mean we have to add some code now which accepts the names regardless of the feature gate *for existing objects* but not for *new objects*, then once all supported releases have that code, allow setting them in new objects? How many releases do we have to wait?

Reply:
> But suppose this goes onto 1.29, someone enables the feature gate, creates an object with a new name, and then downgrades. In 1.28 the name was not know and validation fails.

If the gate is alpha and they opt-in and persist data which a 1.28 server considers invalid, that's on them. In the docs for this, state all servers must be 1.29+ before trying out the alpha feature. We would not promote the gate to beta or enable it by default until the oldest skewed server would recognize / tolerate the data.

> Does that mean we have to add some code now which accepts the names regardless of the feature gate for existing objects but not for new objects,

yes, that's what I was trying to describe in my "informed by the feature gate OR call sites validating updates to existing objects already containing this additional resource type" comment, sorry that wasn't clear.

> then once all supported releases have that code, allow setting them in new objects? How many releases do we have to wait?

we only have to wait a single release, so the n-1 server has that code, before toggling on the gate and allowing that data in.

Reply:
> If the gate is alpha and they opt-in and persist data which a 1.28 server considers invalid, that's on them. 

I agree - just wanted to be sure.

> sorry that wasn't clear

It was clear, just not whether downgrading to 1.28 must be supported.

Reply:
@liggitt: can you take another look? I think I have implemented the validation as intended.